stm32 auto-generate project from .ioc (linux only)

buha opened the pull request:
To build a project and run it on any stm32 board (nucleo, sdp-k1, mbed etc.) the user can now do the following:

- install prerequisites (`stm32cubeide`, `stm32cubemx`)
    - `sed`, `head` are also prerequisites but they are already are installed on any linux
    - `java` is also a prerequisite but it's a dependency of `stm32cubemx`
- obtain / create an .ioc file for your board using `stm32cubemx` (or `stm32cubeide`)
    - the .ioc should at least specify the core clock configuration (RCC) settings and enable the used peripherals and their corresponding pins
    - setting more specific peripheral settings (e.g.: baudrate) is not needed, these are specified in no-OS init params and the stm32 no-OS platform drivers take care of the initialization of these settings.
- copy the .ioc to the project directory and build using `make`
- `make run` / `make debug` to run / debug via command line
- `make develop` to launch the generated project visually

Here's a sample .ioc file for SDP-CK1Z board that can be used to run projects/iio_demo already:
`$ cat projects/iio_demo/k1.ioc`
```
#MicroXplorer Configuration settings - do not modify
File.Version=6
I2C1.I2C_Mode=I2C_Fast
I2C1.IPParameters=I2C_Mode
KeepUserPlacement=false
Mcu.Family=STM32F4
Mcu.IP0=I2C1
Mcu.IP1=NVIC
Mcu.IP2=RCC
Mcu.IP3=SYS
Mcu.IP4=UART5
Mcu.IPNb=5
Mcu.Name=STM32F469NIHx
Mcu.Package=TFBGA216
Mcu.Pin0=PB8
Mcu.Pin1=PC12
Mcu.Pin2=PB7
Mcu.Pin3=PD2
Mcu.Pin4=PH0/OSC_IN
Mcu.Pin5=PH1/OSC_OUT
Mcu.Pin6=VP_SYS_VS_Systick
Mcu.PinsNb=7
Mcu.ThirdPartyNb=0
Mcu.UserConstants=
Mcu.UserName=STM32F469NIHx
MxCube.Version=6.3.0
MxDb.Version=DB.6.0.30
NVIC.BusFault_IRQn=true\:0\:0\:false\:false\:true\:false\:false
NVIC.DebugMonitor_IRQn=true\:0\:0\:false\:false\:true\:false\:false
NVIC.ForceEnableDMAVector=true
NVIC.HardFault_IRQn=true\:0\:0\:false\:false\:true\:false\:false
NVIC.MemoryManagement_IRQn=true\:0\:0\:false\:false\:true\:false\:false
NVIC.NonMaskableInt_IRQn=true\:0\:0\:false\:false\:true\:false\:false
NVIC.PendSV_IRQn=true\:0\:0\:false\:false\:true\:false\:false
NVIC.PriorityGroup=NVIC_PRIORITYGROUP_4
NVIC.SVCall_IRQn=true\:0\:0\:false\:false\:true\:false\:false
NVIC.SysTick_IRQn=true\:0\:0\:false\:false\:true\:false\:true
NVIC.UsageFault_IRQn=true\:0\:0\:false\:false\:true\:false\:false
PB7.GPIOParameters=GPIO_Pu
PB7.GPIO_Pu=GPIO_PULLUP
PB7.Locked=true
PB7.Mode=I2C
PB7.Signal=I2C1_SDA
PB8.GPIOParameters=GPIO_Pu
PB8.GPIO_Pu=GPIO_PULLUP
PB8.Mode=I2C
PB8.Signal=I2C1_SCL
PC12.Mode=Asynchronous
PC12.Signal=UART5_TX
PD2.Mode=Asynchronous
PD2.Signal=UART5_RX
PH0/OSC_IN.Mode=HSE-External-Oscillator
PH0/OSC_IN.Signal=RCC_OSC_IN
PH1/OSC_OUT.Mode=HSE-External-Oscillator
PH1/OSC_OUT.Signal=RCC_OSC_OUT
PinOutPanel.CurrentBGAView=Top
PinOutPanel.RotationAngle=0
ProjectManager.AskForMigrate=true
ProjectManager.BackupPrevious=false
ProjectManager.CompilerOptimize=6
ProjectManager.ComputerToolchain=false
ProjectManager.CoupleFile=false
ProjectManager.CustomerFirmwarePackage=
ProjectManager.DefaultFWLocation=true
ProjectManager.DeletePrevious=true
ProjectManager.DeviceId=STM32F469NIHx
ProjectManager.FirmwarePackage=STM32Cube FW_F4 V1.26.2
ProjectManager.FreePins=false
ProjectManager.HalAssertFull=false
ProjectManager.HeapSize=0x200
ProjectManager.KeepUserCode=true
ProjectManager.LastFirmware=true
ProjectManager.LibraryCopy=1
ProjectManager.MainLocation=Core/Src
ProjectManager.NoMain=false
ProjectManager.PreviousToolchain=
ProjectManager.ProjectBuild=false
ProjectManager.ProjectFileName=k1.ioc
ProjectManager.ProjectName=k1
ProjectManager.RegisterCallBack=
ProjectManager.StackSize=0x400
ProjectManager.TargetToolchain=STM32CubeIDE
ProjectManager.ToolChainLocation=
ProjectManager.UnderRoot=true
ProjectManager.functionlistsort=1-MX_GPIO_Init-GPIO-false-HAL-true,2-SystemClock_Config-RCC-false-HAL-false,3-MX_I2C1_Init-I2C1-false-HAL-true,4-MX_USART2_UART_Init-USART2-false-HAL-true
RCC.AHBFreq_Value=24000000
RCC.APB1Freq_Value=24000000
RCC.APB1TimFreq_Value=24000000
RCC.APB2Freq_Value=24000000
RCC.APB2TimFreq_Value=24000000
RCC.CortexFreq_Value=24000000
RCC.DSIFreq_Value=20000000
RCC.DSITXEscFreq_Value=5000000
RCC.EthernetFreq_Value=24000000
RCC.FCLKCortexFreq_Value=24000000
RCC.FamilyName=M
RCC.HCLKFreq_Value=24000000
RCC.HSE_VALUE=8000000
RCC.HSICalibrationValue=16
RCC.I2SFreq_Value=96000000
RCC.IPParameters=AHBFreq_Value,APB1Freq_Value,APB1TimFreq_Value,APB2Freq_Value,APB2TimFreq_Value,CortexFreq_Value,DSIFreq_Value,DSITXEscFreq_Value,EthernetFreq_Value,FCLKCortexFreq_Value,FamilyName,HCLKFreq_Value,HSE_VALUE,HSICalibrationValue,I2SFreq_Value,LCDTFTFreq_Value,MCO2PinFreq_Value,PLLCLKFreq_Value,PLLDSIFreq_Value,PLLDSIVCOFreq_Value,PLLI2SQCLKFreq_Value,PLLI2SRCLKFreq_Value,PLLM,PLLP,PLLQCLKFreq_Value,PLLRCLKFreq_Value,PLLRFreq_Value,PLLSAIPCLKFreq_Value,PLLSAIQCLKFreq_Value,PLLSAIRCLKFreq_Value,PLLSourceVirtual,RTCFreq_Value,RTCHSEDivFreq_Value,SAIAFreq_Value,SAIBFreq_Value,SDIOFreq_Value,SYSCLKFreq_VALUE,SYSCLKSource,USBFreq_Value,VCOI2SOutputFreq_Value,VCOInputFreq_Value,VCOOutputFreq_Value,VCOSAIOutputFreq_Value
RCC.LCDTFTFreq_Value=48000000
RCC.MCO2PinFreq_Value=24000000
RCC.PLLCLKFreq_Value=24000000
RCC.PLLDSIFreq_Value=160000000
RCC.PLLDSIVCOFreq_Value=320000000
RCC.PLLI2SQCLKFreq_Value=48000000
RCC.PLLI2SRCLKFreq_Value=96000000
RCC.PLLM=8
RCC.PLLP=RCC_PLLP_DIV8
RCC.PLLQCLKFreq_Value=48000000
RCC.PLLRCLKFreq_Value=96000000
RCC.PLLRFreq_Value=96000000
RCC.PLLSAIPCLKFreq_Value=96000000
RCC.PLLSAIQCLKFreq_Value=48000000
RCC.PLLSAIRCLKFreq_Value=96000000
RCC.PLLSourceVirtual=RCC_PLLSOURCE_HSE
RCC.RTCFreq_Value=32000
RCC.RTCHSEDivFreq_Value=4000000
RCC.SAIAFreq_Value=48000000
RCC.SAIBFreq_Value=48000000
RCC.SDIOFreq_Value=48000000
RCC.SYSCLKFreq_VALUE=24000000
RCC.SYSCLKSource=RCC_SYSCLKSOURCE_PLLCLK
RCC.USBFreq_Value=48000000
RCC.VCOI2SOutputFreq_Value=192000000
RCC.VCOInputFreq_Value=1000000
RCC.VCOOutputFreq_Value=192000000
RCC.VCOSAIOutputFreq_Value=192000000
UART5.IPParameters=VirtualMode
UART5.VirtualMode=Asynchronous
VP_SYS_VS_Systick.Mode=SysTick
VP_SYS_VS_Systick.Signal=SYS_VS_Systick
board=custom
isbadioc=false
```

buha commented on tools/scripts/generic.mk:
```diff
@@ -90,7 +90,6 @@ get_relative_path = $(RELATIVE_PATH)
 get_full_path = $(FULL_PATH)
 endif
 
-relative_to_project = $(addprefix $(PROJECT_BUILD)/,$(call get_relative_path,$1))
```
need to do this for all platforms, not just aducm and stm32, will fix in V2

buha commented on tools/scripts/stm32.mk:
```diff
@@ -1,10 +1,94 @@
+# Get the path of the .s script
+ASM_SRCS += $(call rwildcard, $(PROJECT_BUILD)/Core/Startup,*.s)
+
+TARGET = $(shell sed -rn 's|^.*(STM32[A-Z][0-9][0-9][0-9]x+)"/>$$|\1|p' /tmp/.cproject | head -n 1)
```
hardcoded path, will fix in V2

buha commented:
V2:
- Fixed my own comments above, this should hopefully allow tests to pass on xilinx/altera/linux

mchindri commented on tools/scripts/generic.mk:
```diff
@@ -221,6 +230,7 @@ SRC_DIRS := $(patsubst %/,%,$(SRC_DIRS))
 SRCS     += $(foreach dir, $(SRC_DIRS), $(call rwildcard, $(dir),*.c))
 INCS     += $(foreach dir, $(SRC_DIRS), $(call rwildcard, $(dir),*.h))
 
+
```
nitpick: extra line

mchindri commented on tools/scripts/stm32.mk:
```diff
@@ -1,10 +1,92 @@
+PLATFORM_FULL_PATH = $1
+
+# stm32 specific implementation of this function, sources need to go under Core directory
+relative_to_project = $(addprefix $(PROJECT_BUILD)/Core/, $(call get_relative_path, $1))
```
Nitpick: Wouldn't it work with BUILD_DIR instead of PROJECT_BUILD. This way relative_to_project won't be duplicated in each file

buha replied:
good point, i will investigate, i don't like duplicating either

mchindri commented on tools/scripts/generic.mk:
```diff
@@ -45,8 +45,8 @@ remove_fun = rm -rf $1
 read_file = cat $1 2> /dev/null
-make_dir_link = ln -s $1 $2
-make_link = ln -s $1 $2
+make_dir_link = ln -sf $1 $2
```
When this may happen?

buha replied:
i think `make update_srcs` does that

buha replied:
so i dropped this commit in v3

mchindri reviewed (commented):
Just 2 questions, otherwise looks good.

buha commented on tools/scripts/stm32.mk:
```diff
@@ -89,18 +151,36 @@ $(BINARY).gdb:
-stm32_run: all openocd_paths $(BINARY).openocd 
-	$(OPENOCD_BIN)/openocd -s "$(OPENOCD_SCRIPTS)" -f $(BINARY).openocd \
+$(PLATFORM)_run: all $(BINARY).openocd
+	openocd -s "$(OPENOCD_SCRIPTS)" -f $(BINARY).openocd \
```
use stm32cubeide toolchain openocd, not the stock one

buha commented:
V3:
- dropped the ln -sf commit
- moved the relative_to_project in generic.mk, after the [platform].mk are included
- PLATFORM_BUILD is now defined in each [PLATFORM].mk due to the fact that there may be differences (stm32 is different than others)
- use toolchain openocd for make run

mchindri commented on tools/scripts/generic.mk:
```diff
@@ -204,6 +211,7 @@ CFLAGS += -Wall								\
 #------------------------------------------------------------------------------
 #                          COMMON INITIALIZATION
 #------------------------------------------------------------------------------
+relative_to_project = $(addprefix $(PROJECT_BUILD)/,$(call get_relative_path,$1))
```
nitpick: It seems after update you didn't put it in the same place

buha replied:
i put it intentionately to this place because I'm pretty sure it needs PROJECT_BUILD to be defined
and PROJECT_BUILD is now defined in [platform].mk just above where I placed it

mchindri replied:
It seems better here then.
Anyway, as a note. If `:=` is used (Ex 'var2 := $(VAR1)`)  then $(VAR1) is evaluated otherwise (with just `=`, just the reference is used and it is evaluated later when actually needed.

mchindri commented on projects/iio_demo/src.mk:
```diff
@@ -62,31 +61,37 @@ endif
+INCS += $(PLATFORM_DRIVERS)/stm32_uart_stdio.h \
+	$(PLATFORM_DRIVERS)/stm32_uart.h \
+	$(PLATFORM_DRIVERS)/stm32_hal.h
+# SRCS += $(STM32CUBE)/STM32CubeF4/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal.c \
```
Is there a reason to keep them as comment?

buha replied:
i forgot the comments, removed in v4

buha commented:
V4:
- remove unnecessary comments in src.mk

adrimbarean commented on drivers/platform/stm32/stm32_hal.h:
```diff
@@ -0,0 +1,7 @@
+#ifndef _STM32_HAL_H
```
What is the purpose of this file?

buha replied:
- it includes an auto-generated file, main.h which in turn includes the chip specific stm32f4xx_hal.h file (dependant on the MCU family)
- it exposes the SystemClock_Config() function which is also autogenerated in a main.c file and is not otherwise exposed. This SystemClock_Config() function is very important, it initializes the core clock, prescalers, PLLs etc. according to what the user set in the .ioc file.
